### ISSUE TITLE
AOT autograd: Avoid dependency on strides for manual regeneration of outputs that are aliased to inputs

### ISSUE THREAD

**Issue by karthiknagasub** [pytorch/pytorch#103650]
### 🐛 Describe the bug

AOT autograd's manual regeneration of outputs assumes strided output tensor  support. But backends like Habana (https://docs.habana.ai/en/latest/index.html) produces contiguous tensor outputs. As a result, examples like the below one results in accuracy issues

 ```
def fn(a):
  b = a.t()
  b.mul_(1.0)
  return b   

x = torch.arange(6).reshape([2, 3]).to('hpu')

print("x ", x.cpu())

compiled_fn = torch.compile(fn, backend="aot_hpu_training_backend")        
y = compiled_fn(x)

print("y ", y.cpu())

```

**Console prints:**

x  tensor([[0, 1, 2],
        [3, 4, 5]])
		
y  tensor([[0, 1],
        [2, 3],
        [4, 5]])
		
Tensor y has incorrect values. Expected output is 
y  tensor([[0, 3],
        [1, 4],
        [2, 5]])
		
Our analysis suggests that the issue occurs in step 4 of runtime_wrapper(), inside gen_alias_from_base(). This function assumes strided tensor for mul's output. Related bug for TPU backend: https://github.com/pytorch/xla/issues/5179

**Request:** 
Please avoid dependency on the tensor strides for output regeneration. Instead the information regarding the original alias operation (transpose in the above example) should be used. 

### Versions

Collecting environment information...
PyTorch version: 2.0.1a0+git7f346ef
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 14.0.5 (ssh://gerrit:29418/tpc_llvm10 38f3bfdbf5b8581125aa7ac5a2605c9674624566)
CMake version: version 3.26.3
Libc version: glibc-2.31

Python version: 3.8.11 (default, Aug  3 2021, 15:09:35)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-137-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

cc @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste @ezyang @msaroufim @anijain2305 @zou3519 @wconstab

**Comment by ezyang** [comment]
@albanD this is another use case for "view_fn but without holding on to the original tensor"

**Comment by mlazos** [comment]
arbitrarily assigned to @Chillee

**Comment by ezyang** [comment]
@karthiknagasub could you give us more details?

**Comment by karthiknagasub** [comment]
Sure. The fx graph after functionalization is as below:
graph():
    %arg0_1 : [#users=1] = placeholder[target=arg0_1]
    %t : [#users=1] = call_function[target=torch.ops.aten.t.default](args = (%arg0_1,), kwargs = {})
    %mul : [#users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%t, 1.0), kwargs = {})
    %_to_copy : [#users=1] = call_function[target=torch.ops.aten._to_copy.default](args = (%mul,), kwargs = {dtype: torch.int64, layout: torch.strided})
    %t_1 : [#users=2] = call_function[target=torch.ops.aten.t.default](args = (%_to_copy,), kwargs = {})
    %t_2 : [#users=1] = call_function[target=torch.ops.aten.t.default](args = (%t_1,), kwargs = {})
    return (t_1, t_2)
    
  Now in backends like Habana , the %mul  will have contiguous strides. This behavior is different from that of CPU. 
 We guess it is applicable for TPU XLA as well as can be seen from below example:

`import torch
import torch_xla

def fn(a):
  b = a.t()
  c = b.mul(1.0)
  return c

compiled_fn = torch.compile(fn, backend="torchxla_trace_once")

x = torch.randn([2, 3])
res = fn(x)
print(f"CPU: input stride = {x.stride()}")
print(f"CPU: result stride = {res.stride()}")
print(f"CPU: result = {res}")

hx = x.to('xla')
hres = compiled_fn(hx)
print(f"XLA: input stride = {hx.stride()}")
print(f"XLA: result stride = {hres.stride()}")
print(f"XLA: result = {hres.cpu()}")`

**Console prints:**
CPU: input stride = (3, 1)
**CPU: result stride = (1, 3)**
CPU: result = tensor([[-0.2313, -1.1834],
        [-0.6903,  0.3262],
        [ 0.9842,  0.1656]])
XLA: input stride = (3, 1)
**XLA: result stride = (1,)**
XLA: result = tensor([[-0.2313, -1.1834],
        [-0.6903,  0.3262],
        [ 0.9842,  0.1656]])
        
Even though the results match between TPU and CPU, the strides are different. 
The request is not to rely on output tensor strides for AOT functionalization. Instead if the original views ops (transpose in the original example) are dispatched with the updated base as input, we believe that all the backends will have correct output. Please let me know if you need additional details

**Comment by ezyang** [comment]
Oh, this is a bit of a strange thing to do. If you have implemented your function (mul) in a way that produces different strides, you have to instruct our meta formulas that your strides are different. Unfortunately, while there is some precedent for doing this with CPU/CUDA (as these sometimes disagree on striding), we have no extension point for overriding on a per device basis.

If you are willing to edit PyTorch, you need to edit fake_tensor.py make_fast_binary_impl to change its stride compute, for mul and other pointwise ops. Other ops will have meta registrations scattered throughout pytorch. I don't have a better story.

cc @zou3519

**Comment by karthiknagasub** [comment]
@ezyang Thanks. HPU backend in compile mode (graph mode in general) produces contiguous tensors for all non-view ops. Can this observation somehow be used to localize the device specific changes? For instance, can we somehow make Faketensor creation to have contiguous strides for all non-view ops?

**Comment by zou3519** [comment]
I have two ideas for a better long-term story, but no suggestions for what to do for the short-term aside from modifying pytorch.

- As ezyang mentioned, Per-backend FakeTensor registrations
- Maybe we should have some sort of stride-agnostic tracing mode (for both Dynamo and AOTAutograd). If the number of backends increases, asking them to model strides correctly is tricky, because it's difficult for us to even describe the stride semantics on many of our operators.

**Comment by ezyang** [comment]
Yeah, stride agnostic pytorch will help (in the sense that we don't need accurate strides anymore.)

**Comment by karthiknagasub** [comment]
@ezyang @zou3519 Could you please review https://github.com/pytorch/pytorch/pull/104689?

**Comment by wuxun-zhang** [comment]
@ezyang Hi, I still see the issue here even with #104689 merged. I suspect it's related to the changes of https://github.com/pytorch/pytorch/pull/102767. Could you please help take another look? Thanks.

BTW, I noticed a similar issue in `permute + index_add_` in Inductor, see https://github.com/pytorch/pytorch/issues/111133.

**Comment by ezyang** [comment]
@wuxun-zhang You'll have to give more details, maybe file an issue of your own with repro instructions?

**Comment by wuxun-zhang** [comment]
Hi @ezyang, thanks for reminder. Here it's what I met: https://github.com/pytorch/pytorch/issues/111417
